File: flask-backend/app/routes/team.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from ..models import Team, TeamMember, Invitation, User, db
from datetime import datetime, timedelta
import uuid

team_bp = Blueprint('team', __name__)

def check_team_permissions(user_id, team_id, required_roles=None):
    """Check if user has required permissions for team operations"""
    member = TeamMember.query.filter_by(
        team_id=team_id,
        user_id=user_id
    ).first()
    
    if not member:
        return False
        
    if required_roles and member.role not in required_roles:
        return False
        
    return True

@team_bp.route('/team/create', methods=['POST'])
@jwt_required()
def create_team():
    try:
        data = request.get_json()
        user_id = get_jwt_identity()
        
        # First create and commit the team
        team = Team(
            name=data['name'],
            owner_id=user_id
        )
        db.session.add(team)
        db.session.flush()  # This gets us the team.id without committing
        
        # Now create the team member with the team_id
        member = TeamMember(
            team_id=team.id,
            user_id=user_id,
            role='owner',
            joined_at=datetime.utcnow()
        )
        db.session.add(member)
        db.session.commit()
        
        return jsonify({
            'team_id': team.id,
            'name': team.name,
            'role': 'owner'
        }), 201
        
    except Exception as e:
        db.session.rollback()
        print(f"Error creating team: {str(e)}")
        return jsonify({'error': 'Failed to create team'}), 500

@team_bp.route('/team/invite', methods=['POST'])
@jwt_required()
def invite_member():
    data = request.get_json()
    user_id = get_jwt_identity()
    
    # Verify user has permission to invite
    team_member = TeamMember.query.filter_by(
        team_id=data['team_id'],
        user_id=user_id
    ).first()
    
    if not team_member or team_member.role == 'member':
        return jsonify({'error': 'Unauthorized'}), 403
    
    invitation = Invitation(
        team_id=data['team_id'],
        email=data['email'],
        role=data['role'],
        invited_by=user_id,
        expires_at=datetime.utcnow() + timedelta(days=7)
    )
    
    db.session.add(invitation)
    db.session.commit()
    
    # Get team name and send invitation email
    team = Team.query.get(data['team_id'])
    invite_url = f"http://localhost:3000/invite/{invitation.invite_code}"
    
    from ..services.email_service import send_team_invitation
    email_sent = send_team_invitation(
        data['email'],
        team.name,
        data['role'],
        invite_url
    )
    
    if not email_sent:
        return jsonify({'error': 'Failed to send invitation email'}), 500
    
    return jsonify({
        'invite_code': invitation.invite_code,
        'invite_url': invite_url
    }), 201

@team_bp.route('/team/accept-invite/<invite_code>', methods=['POST'])
@jwt_required()
def accept_invite(invite_code):
    user_id = get_jwt_identity()
    invitation = Invitation.query.filter_by(
        invite_code=invite_code,
        status='pending'
    ).first()
    
    if not invitation or invitation.expires_at < datetime.utcnow():
        return jsonify({'error': 'Invalid or expired invitation'}), 400
    
    member = TeamMember(
        team_id=invitation.team_id,
        user_id=user_id,
        role=invitation.role,
        invited_by=invitation.invited_by
    )
    
    invitation.status = 'accepted'
    db.session.add(member)
    db.session.commit()
    
    return jsonify({'message': 'Invitation accepted successfully'}), 200

@team_bp.route('/team/members/<team_id>', methods=['GET'])
@jwt_required()
def get_team_members(team_id):
    user_id = get_jwt_identity()
    
    # Check if user is part of the team
    if not TeamMember.query.filter_by(team_id=team_id, user_id=user_id).first():
        return jsonify({'error': 'Unauthorized'}), 403
    
    members = TeamMember.query.filter_by(team_id=team_id).all()
    members_data = []
    
    for member in members:
        user = User.query.get(member.user_id)
        members_data.append({
            'id': member.id,
            'user_id': member.user_id,
            'email': user.email,
            'role': member.role,
            'joined_at': member.joined_at.isoformat()
        })
    
    return jsonify({'members': members_data}), 200

File: express-backend/src/routes/team.js
const express = require('express');
const router = express.Router();
const { Team, TeamMember, Invitation, User, ActivityLog } = require('../models');
const { authMiddleware, checkTeamPermissions, requireTeamRole } = require('../middleware/auth');
const { validate, schemas } = require('../middleware/validation');
const emailService = require('../services/emailService');
const { Op } = require('sequelize');

// Create team route
router.post('/create', authMiddleware, validate(schemas.createTeam), async (req, res) => {
    try {
        const { name } = req.body;
        const userId = req.user.id;

        const team = await Team.create({
            name,
            owner_id: userId
        });

        await TeamMember.create({
            team_id: team.id,
            user_id: userId,
            role: 'owner'
        });

        // Log activity
        await ActivityLog.logActivity(
            team.id,
            userId,
            'team_created',
            'team',
            team.id,
            { team_name: name }
        );

        res.status(201).json({
            team_id: team.id,
            name: team.name,
            role: 'owner'
        });
    } catch (error) {
        console.error('Error creating team:', error);
        res.status(500).json({ error: error.message });
    }
});

// Invite member route - FIXED
router.post('/invite', authMiddleware, validate(schemas.inviteMember), async (req, res) => {
    try {
        const { team_id, email, role } = req.body;
        const userId = req.user.id;

        // Check permissions - FIXED: Properly check for owner or admin
        const hasPermission = await checkTeamPermissions(userId, team_id, ['owner', 'admin']);

        if (!hasPermission) {
            return res.status(403).json({ error: 'Unauthorized - Only owners and admins can invite members' });
        }

        // Check if user is already a member
        const existingMember = await TeamMember.findOne({
            include: [{
                model: User,
                where: { email }
            }],
            where: { team_id }
        });

        if (existingMember) {
            return res.status(400).json({ error: 'User is already a team member' });
        }

        // Check for pending invitation
        const existingInvitation = await Invitation.findOne({
            where: {
                team_id,
                email,
                status: 'pending',
                expires_at: { [Op.gt]: new Date() }
            }
        });

        if (existingInvitation) {
            return res.status(400).json({ error: 'Invitation already sent to this email' });
        }

        // Create invitation
        const invitation = await Invitation.create({
            team_id,
            email,
            role,
            invited_by: userId,
            expires_at: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000) // 7 days
        });

        // Get team name
        const team = await Team.findByPk(team_id);

        // Use environment variable for frontend URL
        const frontendUrl = process.env.FRONTEND_URL || 'http://localhost:3000';
        const invite_url = `${frontendUrl}/auth/team/${invitation.invite_code}`;

        // Send email
        const emailSent = await emailService.sendTeamInvitation(
            email,
            team.name,
            role,
            invite_url
        );

        if (!emailSent) {
            console.warn('Failed to send invitation email');
        }

        res.status(201).json({
            invite_code: invitation.invite_code,
            invite_url: invite_url,
            email_sent: emailSent
        });
    } catch (error) {
        console.error('Error creating invitation:', error);
        res.status(500).json({ error: error.message });
    }
});

// Accept invitation route
router.post('/accept-invite/:invite_code', authMiddleware, async (req, res) => {
    try {
        const { invite_code } = req.params;
        const userId = req.user.id;

        const invitation = await Invitation.findOne({
            where: {
                invite_code,
                status: 'pending'
            }
        });

        if (!invitation) {
            return res.status(404).json({ error: 'Invitation not found' });
        }

        if (invitation.expires_at < new Date()) {
            invitation.status = 'expired';
            await invitation.save();
            return res.status(400).json({ error: 'Invitation has expired' });
        }

        // Check if already a member
        const existingMember = await TeamMember.findOne({
            where: {
                team_id: invitation.team_id,
                user_id: userId
            }
        });

        if (existingMember) {
            return res.status(400).json({ error: 'You are already a member of this team' });
        }

        await TeamMember.create({
            team_id: invitation.team_id,
            user_id: userId,
            role: invitation.role,
            invited_by: invitation.invited_by
        });

        invitation.status = 'accepted';
        await invitation.save();

        res.json({ message: 'Invitation accepted successfully' });
    } catch (error) {
        console.error('Error accepting invitation:', error);
        res.status(500).json({ error: error.message });
    }
});

// Get team members route
router.get('/members/:team_id', authMiddleware, async (req, res) => {
    try {
        const { team_id } = req.params;
        const userId = req.user.id;

        const hasPermission = await checkTeamPermissions(userId, team_id);
        if (!hasPermission) {
            return res.status(403).json({ error: 'Unauthorized' });
        }

        const members = await TeamMember.findAll({
            where: { team_id },
            include: [{
                model: User,
                attributes: ['id', 'email', 'status']
            }],
            order: [['joined_at', 'DESC']]
        });

        const membersData = members.map(member => ({
            id: member.id,
            user_id: member.user_id,
            email: member.User.email,
            role: member.role,
            joined_at: member.joined_at
        }));

        res.json({ members: membersData });
    } catch (error) {
        console.error('Error fetching team members:', error);
        res.status(500).json({ error: error.message });
    }
});

// Remove team member (owner/admin only)
router.delete('/members/:team_id/:user_id', authMiddleware, async (req, res) => {
    try {
        const { team_id, user_id } = req.params;
        const requesterId = req.user.id;

        // Check if requester has permission
        const hasPermission = await checkTeamPermissions(requesterId, team_id, ['owner', 'admin']);
        if (!hasPermission) {
            return res.status(403).json({ error: 'Unauthorized - Only owners and admins can remove members' });
        }

        // Check if target user is the owner
        const team = await Team.findByPk(team_id);
        if (team.owner_id === user_id) {
            return res.status(400).json({ error: 'Cannot remove the team owner' });
        }

        // Check if requester is trying to remove themselves
        if (requesterId === user_id) {
            return res.status(400).json({ error: 'Use the leave team endpoint to remove yourself' });
        }

        const deleted = await TeamMember.destroy({
            where: {
                team_id,
                user_id
            }
        });

        if (deleted === 0) {
            return res.status(404).json({ error: 'Team member not found' });
        }

        res.json({ message: 'Member removed successfully' });
    } catch (error) {
        console.error('Error removing team member:', error);
        res.status(500).json({ error: error.message });
    }
});

// Leave team
router.post('/leave/:team_id', authMiddleware, async (req, res) => {
    try {
        const { team_id } = req.params;
        const userId = req.user.id;

        // Check if user is the owner
        const team = await Team.findByPk(team_id);
        if (!team) {
            return res.status(404).json({ error: 'Team not found' });
        }

        if (team.owner_id === userId) {
            return res.status(400).json({
                error: 'Team owner cannot leave the team. Transfer ownership or delete the team instead.'
            });
        }

        const deleted = await TeamMember.destroy({
            where: {
                team_id,
                user_id: userId
            }
        });

        if (deleted === 0) {
            return res.status(404).json({ error: 'You are not a member of this team' });
        }

        res.json({ message: 'Successfully left the team' });
    } catch (error) {
        console.error('Error leaving team:', error);
        res.status(500).json({ error: error.message });
    }
});

module.exports = router;
